lib/Frontend/ParseableInterfaceSupport.cpp

```diff
@@ -194,34 +188,83 @@ swiftModuleIsUpToDate(clang::vfs::FileSystem &FS,
+///
+///    - For any file in the latter set that is itself a .swiftmodule
+///      living in \p ModuleCachePath, all of _its_ dependencies, copied
+///      out to avoid having to do recursive scanning when rechecking this
+///      dependency in the future.
+static bool
+collectDepsForSerialization(clang::vfs::FileSystem &FS,
+                            CompilerInstance &SubInstance,
+                            StringRef InPath, StringRef ModuleCachePath,
+                            SmallVectorImpl<FileDependency> &Deps,
+                            DiagnosticEngine &Diags) {
+  auto DTDeps = SubInstance.getDependencyTracker()->getDependencies();
+  SmallVector<StringRef, 16> InitialDepNames(DTDeps.begin(), DTDeps.end());
+  InitialDepNames.push_back(InPath);
+  llvm::StringSet<> AllDepNames;
+  for (auto const &DepName : InitialDepNames) {
+    AllDepNames.insert(DepName);
+    auto DepBuf = getBufferOfDependency(FS, InPath, DepName, Diags);
+    if (!DepBuf) {
+      return true;
+    }
+    uint64_t Size = DepBuf->getBufferSize();
+    uint64_t Hash = xxHash64(DepBuf->getBuffer());
+    Deps.push_back(FileDependency{Size, Hash, DepName});
+
+    // If Dep is itself a .swiftmodule in the cache dir, pull out its deps
+    // and include them in our own, so we have a single-file view of
+    // transitive deps: removes redundancies, and avoids opening and reading
+    // multiple swiftmodules during future loads.
+    auto Ext = llvm::sys::path::extension(DepName);
     auto Ty = file_types::lookupTypeForExtension(Ext);
     if (Ty == file_types::TY_SwiftModuleFile &&
-        In.Path.startswith(ModuleCachePath) &&
-        !swiftModuleIsUpToDate(FS, ModuleCachePath, In.Path, Diags)) {
-      LLVM_DEBUG(llvm::dbgs() << "Dep " << In.Path
-                 << " is indirectly out of date\n");
-      return false;
+        DepName.startswith(ModuleCachePath)) {
+      SmallVector<FileDependency, 16> SubDeps;
+      auto VI = serialization::validateSerializedAST(
+          DepBuf->getBuffer(),
+          /*ExtendedValidationInfo=*/nullptr, &SubDeps);
+      if (VI.status != serialization::Status::Valid) {
+        Diags.diagnose(SourceLoc(),
+                       diag::error_extracting_dependencies_from_cached_module,
+                       DepName);
+        return true;
+      }
+      for (auto const &SubDep : SubDeps) {
+        if (AllDepNames.count(SubDep.Path) == 0) {
+          AllDepNames.insert(SubDep.Path);
+          Deps.push_back(SubDep);
+        }
+      }
     }
-    LLVM_DEBUG(llvm::dbgs() << "Dep " << In.Path << " is up to date\n");
   }
-  return true;
+  return false;
 }
 
 static bool buildSwiftModuleFromSwiftInterface(
     clang::vfs::FileSystem &FS, DiagnosticEngine &Diags,
-    CompilerInvocation &SubInvocation, StringRef InPath, StringRef OutPath) {
+    CompilerInvocation &SubInvocation, StringRef InPath, StringRef OutPath,
+                                               StringRef ModuleCachePath) {
```

Comment:
Weird (auto-)indentation here.

Reply:
This looks more like an Xcode-ism.

Reply:
Emacs if anything! But it gets resolved in 40f30cee606ee24359b423aa1119351eb477ba2d anyway.